main.py

```diff
@@ -176,30 +207,109 @@ def clicked(self, position):
-                        ball.select(False)
-                    self.isBallSelected = False
-                # Case(3): A ball is selected atm and the observed ball has a small rim.
-                #   -> Swap the colours of the selected and the observed ball, unselect
-                #      all balls and set depending variables accordingly.
-                elif self.isBallSelected and ball.selected == BALL_MARK_S:
-                    ball.colour = self.selectedBall.colour
-                    self.selectedBall.colour = WHITE
-                    for ball in self.balls:
-                        ball.select(False)
-                    self.isBallSelected = False
-                    self.counter += 1
+                self.move(ball)
                 break
 
+    def setState(self, state):
+        """
+        Function Game.setState(self, state):
+        Sets the state of the current game aka loading a game.
+
+        parma[in]   state   state to set / load
+        """
+        for i in range(len(self.balls)):
+            self.balls[i].colour = state[i]
+            self.balls[i].selected = BALL_NOMARK
+            self.isBallSelected = False
+
+    def getState(self):
+        """
+        Function Game.getState(self):
+        Gets the state of the current game aka saving a game.
+
+        return  state   state to get / safe
+        """
+        state = []
+        for i in range(len(self.balls)):
+            state.append(self.balls[i].colour)
+        return state
+
+class State:
+    def __init__(self, id, inputState, parent, moves):
+        """
+        Function State.__init__(self, id, inputState, parent, moves):
+        This class defines the states relation by adding a parent and used moves to 
+        the state constellation. Also neighbours can be calculated.
+
+        parma[in]   id          state id
+        parma[in]   inputState  real state data
+        parma[in]   parent      state parent according to the shortest route
+        parma[in]   moves       moves needed to reach this state from START
+        """
+        self.id = id
+        self.state = inputState
+        self.parent = parent
+        self.distanceToStart = moves
+        self.distanceToGoal = 0
+        self.stateValue = 0
+        self.calculateDistance()
+
+    def calculateDistance(self):
+        """
+        Function State.calculateDistance(self):
+        Calculate the distance of the current state to the GOAL state
+        by assuming the player can move every ball everywhere. At least
+        one white "ball" has to be moved otherwise switching to balls
+        would only cost one move.
+        """
+        self.distanceToGoal = 0
+        noWhite = True
+        for index in range(len(self.state)):
+            if self.state[index] != GOAL[index]:
+                self.distanceToGoal += 1
+                if self.state[index] == 0:
+                    noWhite = False
+        if noWhite:
+            self.distanceToGoal += 1
```

Comment:
Yes it does for the current way the distance function is used. An example is the distance comparions between START and GOAL. For the first and only level so far, all 8 balls have to be moved. But the distance have to be 9, since two balls can't just switch places. So there has to be an extra field used for all where the first ball can move to ignoring all the rules like "you can't jump over anyone". This is considered the extra move which is checked with noWhite.